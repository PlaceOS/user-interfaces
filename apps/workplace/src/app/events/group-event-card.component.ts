import { Component, Input } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { SettingsService } from '@placeos/common';
import { GroupEventDetailsModalComponent } from '@placeos/bookings';

@Component({
    selector: 'group-event-card',
    template: `
        <button
            matRipple
            (click)="viewDetails()"
            *ngIf="!featured; else featured_card"
            class="border border-base-300 hover:border-info flex flex-col bg-base-100 rounded-xl shadow hover:shadow-2xl overflow-hidden w-60 h-[20rem]"
        >
            <div
                class="flex items-center justify-between h-28 w-full bg-base-200 overflow-hidden"
            >
                <img
                    *ngIf="event.images?.length"
                    auth
                    [source]="event.images[0]"
                    class="object-cover"
                />
            </div>
            <div class="p-4 flex-1 h-1/2">
                <div class="opacity-60 text-sm text-left">
                    {{ event.date | date: 'EEE d MMM' }},
                    {{ event.date | date: time_format }}
                </div>
                <h2
                    class="text-xl mb-2 text-left truncate"
                    [title]="event.title"
                >
                    {{ event.title }}
                </h2>
                <div
                    class="opacity-60 text-xs flex-1 overflow-hidden h-[4.5rem] mb-2"
                >
                    {{ event.description }}
                </div>
                <div class="flex items-center space-x-2 text-sm">
                    <app-icon class="text-info">place</app-icon>
                    <div class="" *ngIf="item?.linked_event?.system_id">
                        {{
                            (item?.linked_event?.system_id | space | async)
                                ?.display_name
                        }}
                    </div>
                    <div
                        class="opacity-30"
                        *ngIf="!item?.linked_event?.system_id"
                    >
                        Remote event
                    </div>
                </div>
                <div class="flex items-center space-x-2 text-sm">
                    <app-icon class="text-info">people</app-icon>
                    <div class="">
                        {{ event.attendees?.length || '0' }} attending
                    </div>
                </div>
            </div>
        </button>
        <ng-template #featured_card>
            <button
                matRipple
                (click)="viewDetails()"
                class="border border-base-300 hover:border-info flex bg-base-100 rounded-xl shadow hover:shadow-2xl overflow-hidden w-[48rem] max-w-full h-56 mx-auto"
            >
                <div
                    class="flex items-center justify-between h-full w-1/2 max-w-[20rem] bg-base-200 overflow-hidden"
                >
                    <img
                        *ngIf="event.images?.length"
                        auth
                        [source]="event.images[0]"
                        class="object-cover"
                    />
                </div>
                <div
                    class="absolute top-0 left-0 rounded-br-xl py-2 pl-2 pr-4 space-x-2 bg-info text-info-content flex items-center text-sm"
                >
                    <app-icon class="text-base">star</app-icon>
                    <div class="uppercase">Featured</div>
                </div>
                <div details class="flex px-8 py-4 space-x-4">
                    <div class="flex flex-col items-center">
                        <div class="text-sm opacity-30 ">
                            {{ event.date | date: 'MMM' }}
                        </div>
                        <div class="text-lg">{{ event.date | date: 'd' }}</div>
                    </div>
                    <div class="flex flex-col space-y-2">
                        <h3 class="text-left">{{ event.title }}</h3>
                        <div time class="text-sm opacity-30">
                            {{ event.date | date: 'EEEE' }}
                            {{ event.date | date: time_format }} -
                            {{
                                event.date + event.duration * 60 * 1000
                                    | date: time_format
                            }}
                        </div>
                        <div class="h-20 overflow-hidden">
                            {{ event.description }}
                        </div>
                        <div class="flex items-center space-x-2 text-sm">
                            <app-icon class="text-info">place</app-icon>
                            <div class="" *ngIf="event.location">
                                {{ event.location }}
                            </div>
                            <div class="opacity-30" *ngIf="!event.location">
                                Remote event
                            </div>
                        </div>
                        <div class="flex items-center space-x-2 text-sm">
                            <app-icon class="text-info">people</app-icon>
                            <div class="">
                                {{ event.attendees?.length || '0' }} attending
                            </div>
                        </div>
                    </div>
                </div>
                <div
                    class="absolute top-4 right-4 bg-secondary text-secondary-content rounded px-4 py-2 w-32 text-center truncate"
                >
                    View Details
                </div>
            </button>
        </ng-template>
    `,
    styles: [
        `
            button {
                transition: box-shadow 300ms, border 200ms;
            }
        `,
    ],
})
export class GroupEventCardComponent {
    @Input() public event: any;
    @Input() public featured: boolean;

    public get time_format(): string {
        return this._settings.time_format;
    }

    constructor(
        private _settings: SettingsService,
        private _dialog: MatDialog
    ) {}

    public ngOnInit(): void {}

    public viewDetails(): void {
        console.log('View Details:', this.event);
        this._dialog.open(GroupEventDetailsModalComponent, {
            data: this.event,
        });
    }
}
