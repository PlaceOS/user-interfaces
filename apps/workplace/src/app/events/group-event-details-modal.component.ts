import { Component, Inject } from '@angular/core';
import { MAT_DIALOG_DATA } from '@angular/material/dialog';
import { Booking } from '@placeos/bookings';
import { SettingsService } from '@placeos/common';

@Component({
    selector: `group-event-details-modal`,
    template: `
        <div
            class="relative w-[48rem] max-w-[calc(100vw-1rem)] max-h-[80vh] overflow-hidden"
        >
            <div
                class="flex items-center justify-between h-52 w-full bg-base-200 overflow-hidden"
            >
                <img
                    *ngIf="booking.images?.length"
                    auth
                    [source]="event.images[0]"
                    class="object-cover"
                />
            </div>
            <div
                class="absolute top-0 left-0 rounded-br py-2 pl-2 pr-4 space-x-2 bg-info text-info-content flex items-center text-sm"
            >
                <app-icon class="text-base">star</app-icon>
                <div class="uppercase">Featured</div>
            </div>
            <button icon mat-dialog-close class="absolute top-1 right-1">
                <app-icon>close</app-icon>
            </button>
            <div
                class="flex items-center justify-between py-4 px-8 border-b border-base-200"
            >
                <h3 class="text-left text-xl">
                    {{ booking.title }}
                </h3>
                <div class="flex items-center space-x-2">
                    <button
                        btn
                        matRipple
                        class="clear bg-base-200 text-base-content space-x-2"
                    >
                        <app-icon>star</app-icon>
                        <div class="pr-2">Interested</div>
                    </button>
                    <button
                        btn
                        matRipple
                        class="clear bg-base-200 text-base-content space-x-2"
                    >
                        <app-icon>help</app-icon>
                        <div class="pr-2">Going</div>
                    </button>
                    <button
                        btn
                        matRipple
                        class="clear bg-base-200 text-base-content w-[2.75rem]"
                    >
                        <app-icon class="text-2xl">more_horiz</app-icon>
                    </button>
                </div>
            </div>
            <div class="flex overflow-y-auto overflow-x-hidden p-8">
                <div class="flex flex-col space-y-2">
                    <div class="flex items-center space-x-4">
                        <div
                            class="flex items-center justify-center w-10 h-10 bg-base-200 rounded-full"
                        >
                            <app-icon>person</app-icon>
                        </div>
                        <div>Event by {{ booking.user_name }}</div>
                    </div>
                    <h3 class="font-medium pt-4">When and where</h3>
                    <div class="flex items-center space-x-4">
                        <div
                            class="flex items-center justify-center w-10 h-10 bg-base-200 rounded-full"
                        >
                            <app-icon>calendar_today</app-icon>
                        </div>
                        <div class="flex flex-col">
                            <div class="text-sm">Date and Time</div>
                            <div class="text-sm opacity-30">
                                {{ booking.date | date: 'EEEE, d MMMM, yyyy' }}
                                . {{ booking.date | date: time_format }} -
                                {{
                                    booking.date + booking.duration * 60 * 1000
                                        | date: time_format
                                }}
                            </div>
                        </div>
                    </div>
                    <div class="flex items-center space-x-4">
                        <div
                            class="flex items-center justify-center w-10 h-10 bg-base-200 rounded-full"
                        >
                            <app-icon>place</app-icon>
                        </div>
                        <div class="text-sm">
                            {{ booking.location }}
                            <span *ngIf="!booking.location" class="opacity-30">
                                Remote Event
                            </span>
                        </div>
                    </div>
                    <div class="flex items-center space-x-4">
                        <div
                            class="flex items-center justify-center w-10 h-10 bg-base-200 rounded-full"
                        >
                            <app-icon>person</app-icon>
                        </div>
                        <div>
                            {{ attendance }} going,
                            {{ booking.attendees?.length - attendance }}
                            interested
                        </div>
                    </div>
                    <h3 class="font-medium pt-4">About this event</h3>
                    <div class="text-sm pb-4">
                        {{ booking.description }}
                        <span *ngIf="!booking.description" class="opacity-30">
                            No description
                        </span>
                    </div>
                </div>
                <div class="flex"></div>
            </div>
        </div>
    `,
    styles: [``],
})
export class GroupEventDetailsModalComponent {
    public booking: any = this._data;

    public get time_format() {
        return this._settings.time_format;
    }

    public get attendance() {
        return (
            this.booking.attendees?.filter((_) => _.status === 'accepted')
                ?.length || 0
        );
    }

    constructor(
        @Inject(MAT_DIALOG_DATA) private _data: Booking,
        private _settings: SettingsService
    ) {}

    public ngOnInit(): void {}
}
