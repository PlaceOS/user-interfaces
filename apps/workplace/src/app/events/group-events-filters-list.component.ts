import { Component } from '@angular/core';
import { GroupEventsStateService } from './group-events-state.service';
import { differenceInDays, endOfDay, startOfDay } from 'date-fns';
import { map, take, tap } from 'rxjs/operators';

@Component({
    selector: 'group-events-filters-list',
    template: `
        <div
            class="my-2 w-[63rem] max-w-full mx-auto border border-base-300 p-4 bg-base-100 rounded"
        >
            <div class="flex items-center justify-between space-x-2 mb-4">
                <div>
                    {{ (this_period | async) ? 'This' : 'Upcoming' }}
                    {{ (period | async) === 'week' ? 'Week' : 'Month' }}
                </div>
            </div>
            <div class="flex flex-wrap">
                <div
                    class="flex items-center px-4 py-3 rounded-3xl border border-base-400 m-1 text-sm"
                >
                    {{ (options | async)?.date | date: 'MMM d, y' }}
                    &ndash;
                    {{ (options | async)?.end | date: 'MMM d, y' }}
                </div>
                <div
                    class="flex items-center pl-4 pr-1 rounded-3xl border border-base-400 m-1"
                    *ngFor="let tag of (filters | async)?.tags || []"
                >
                    <div class="mr-2 flex-1 text-sm">{{ tag }}</div>
                    <button icon matRipple (click)="removeTag(tag)">
                        <app-icon>close</app-icon>
                    </button>
                </div>
            </div>
        </div>
    `,
    styles: [``],
})
export class GroupEventsFiltersListComponent {
    public readonly options = this._state.options;
    public readonly filters = this._state.filters;
    public readonly this_period = this._state.options.pipe(
        map(({ date, end }) => {
            return (
                Date.now() >= startOfDay(date).valueOf() &&
                Date.now() < endOfDay(end || date).valueOf()
            );
        })
    );
    public readonly period = this._state.options.pipe(
        map(({ date, end }) =>
            Math.abs(differenceInDays(date, end || Date.now())) > 7
                ? 'month'
                : 'week'
        )
    );

    constructor(private _state: GroupEventsStateService) {}

    public async removeTag(tag: string) {
        const tags = (await this.filters.pipe(take(1)).toPromise())?.tags || [];
        this._state.setFilters({ tags: tags.filter((_) => _ !== tag) });
    }
}
