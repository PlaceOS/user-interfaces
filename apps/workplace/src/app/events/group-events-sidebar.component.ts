import { Component } from '@angular/core';
import { AsyncHandler, SettingsService } from '@placeos/common';
import {
    addDays,
    addMonths,
    endOfDay,
    format,
    startOfMonth,
    startOfWeek,
} from 'date-fns';
import { BehaviorSubject } from 'rxjs';
import { GroupEventsStateService } from './group-events-state.service';
import { take } from 'rxjs/operators';

@Component({
    selector: `group-events-sidebar`,
    template: `
        <div class="flex flex-col bg-base-100 w-[18rem] h-full">
            <div class="flex items-center space-x-2 p-2">
                <button
                    btn
                    matRipple
                    class="flex-1 rounded-3xl"
                    [class.inverse]="(period | async) !== 'week'"
                    (click)="period.next('week')"
                >
                    Week
                </button>
                <button
                    btn
                    matRipple
                    class="flex-1 rounded-3xl"
                    [class.inverse]="(period | async) !== 'month'"
                    (click)="period.next('month')"
                >
                    Month
                </button>
            </div>
            <div class="flex flex-col items-center space-y-2 pb-2 px-2">
                <mat-form-field
                    appearance="outline"
                    class="w-full no-subscript"
                >
                    <mat-select
                        [(ngModel)]="selected_range"
                        (ngModelChange)="setPeriod($event)"
                        placeholder="Select Period"
                    >
                        <mat-option
                            [value]="range.id"
                            *ngFor="let range of period_list"
                        >
                            {{ range.display }}
                        </mat-option>
                    </mat-select>
                </mat-form-field>
            </div>
            <hr class="border-base-200 w-[calc(100%-1rem)] mx-auto" />
            <date-calendar
                [ngModel]="(options | async).date"
                (ngModelChange)="setPeriodFromDate($event)"
            ></date-calendar>
            <hr class="border-base-200 w-[calc(100%-1rem)] mx-auto" />
            <div class="flex flex-col flex-1 overflow-auto">
                <h2 class="text-lg font-medium p-4">Filters</h2>
                <div
                    class="flex flex-col space-y-2 px-4"
                    *ngIf="(tags | async)?.length"
                >
                    <h3>Tags</h3>
                    <button
                        matRipple
                        class="flex items-center rounded w-full text-left"
                        *ngFor="let tag of tags | async"
                        (click)="toggleTag(tag)"
                    >
                        <mat-checkbox
                            [ngModel]="(filters | async)?.tags?.includes(tag)"
                        >
                            {{ tag }}
                        </mat-checkbox>
                    </button>
                </div>
            </div>
        </div>
    `,
    styles: [``],
})
export class GroupEventsSidebarComponent extends AsyncHandler {
    public period = new BehaviorSubject<'week' | 'month'>('week');
    public period_list = [];
    public selected_range: number;
    public readonly options = this._state.options;
    public readonly filters = this._state.filters;
    public readonly tags = this._state.tags;

    constructor(
        private _settings: SettingsService,
        private _state: GroupEventsStateService
    ) {
        super();
    }

    public ngOnInit(): void {
        this.subscription(
            'period',
            this.period.subscribe(() => {
                this._generatePeriods();
                if (this.period_list.length) {
                    this.setPeriod(this.period_list[0].id);
                    this.selected_range = this.period_list[0].id;
                }
            })
        );
        this._generatePeriods();
        if (this.period_list.length) {
            this.setPeriod(this.period_list[0].id);
            this.selected_range = this.period_list[0].id;
        }
    }

    public async toggleTag(tag: string) {
        const tags = (await this.filters.pipe(take(1)).toPromise())?.tags || [];
        if (tags.includes(tag)) {
            this._state.setFilters({ tags: tags.filter((_) => _ !== tag) });
        } else {
            this._state.setFilters({ tags: [...tags, tag] });
        }
    }

    public setPeriodFromDate(date: number) {
        for (const period of this.period_list) {
            if (date >= period.start && date <= period.end) {
                this.selected_range = period.id;
                this.setPeriod(period.id);
                return;
            }
        }
    }

    public setPeriod(id: string) {
        const { start, end } = this.period_list.find((_) => _.id === id);
        this._state.setOptions({ date: start, end });
    }

    private _generatePeriods() {
        const periods = [];
        const period_type = this.period.value;
        let date = Date.now();
        const end_date = addDays(date, 12 * 30).valueOf();
        const week_offset = this._settings.get('app.week_start') || 0;
        if (period_type === 'month') {
            date = startOfMonth(date).valueOf();
        } else if (period_type === 'week') {
            date = startOfWeek(date, { weekStartsOn: week_offset }).valueOf();
        }
        while (date < end_date) {
            if (period_type === 'week') {
                const end = endOfDay(addDays(date, 6)).valueOf();
                periods.push({
                    id: date,
                    start: date,
                    end,
                    display: `${format(
                        Math.max(Date.now(), date),
                        'EEE, do MMM'
                    )} â€“ ${format(end, 'do MMM')}`,
                });
                date = addDays(date, 7).valueOf();
            } else if (period_type === 'month') {
                const end = endOfDay(addDays(addMonths(date, 1), -1)).valueOf();
                periods.push({
                    id: date,
                    start: date,
                    end,
                    display: `${format(date, 'MMMM yyyy')}`,
                });
                date = addMonths(date, 1).valueOf();
            } else break;
        }
        this.period_list = periods;
    }
}
