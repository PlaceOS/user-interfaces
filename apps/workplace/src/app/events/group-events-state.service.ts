import { Injectable } from '@angular/core';
import { queryBookings } from '@placeos/bookings';
import { SettingsService, unique } from '@placeos/common';
import { OrganisationService } from '@placeos/organisation';
import { endOfDay, getUnixTime, startOfDay } from 'date-fns';
import { BehaviorSubject, combineLatest } from 'rxjs';
import { filter, map, shareReplay, switchMap, tap } from 'rxjs/operators';

export interface GroupEventOptions {
    date: number;
    end?: number;
}

export interface GroupEventFilters {
    categories: string[];
    tags: string[];
}

@Injectable({
    providedIn: 'root',
})
export class GroupEventsStateService {
    private _options = new BehaviorSubject<GroupEventOptions>({
        date: Date.now(),
    });
    private _filters = new BehaviorSubject<GroupEventFilters>({
        categories: [],
        tags: [],
    });
    private _tag_list = new BehaviorSubject<string[]>([]);

    public readonly filters = this._filters.asObservable();
    public readonly tags = this._tag_list.asObservable();

    public readonly events = combineLatest([
        this._org.active_building,
        this._options,
    ]).pipe(
        filter(([building]) => !!building),
        switchMap(([building, options]) =>
            queryBookings({
                period_start: getUnixTime(
                    startOfDay(Math.max(Date.now(), options.date))
                ),
                period_end: getUnixTime(endOfDay(options.end || options.date)),
                type: 'group-event',
                zones: this._settings.get('app.use_region')
                    ? building.parent_id
                    : building.id,
            })
        ),
        map((list) =>
            list
                .filter((_) => _.permission !== 'PRIVATE')
                .sort((a, b) => a.date - b.date)
        ),
        tap((list) => {
            const old_tags = this._tag_list.getValue();
            const tags = list.map((event) => event.tags).flat();
            this._tag_list.next(unique([...old_tags, ...tags]));
        }),
        shareReplay(1)
    );

    public readonly filtered_events = combineLatest([
        this.events,
        this._filters,
    ]).pipe(
        map(([list, { tags }]) => {
            const tag_list = tags.map((_) => _.toLowerCase());
            return list.filter((event) => {
                const event_tags = event.tags.map((_) => _.toLowerCase());
                return (
                    tag_list.every((tag) => event_tags.includes(tag)) &&
                    event.date_end > Date.now()
                );
            });
        }),
        shareReplay(1)
    );

    public readonly options = this._options.asObservable();

    constructor(
        private _org: OrganisationService,
        private _settings: SettingsService
    ) {}

    public setOptions(options: Partial<GroupEventOptions>) {
        this._options.next({ ...this._options.value, ...options });
    }

    public setFilters(filters: Partial<GroupEventFilters>) {
        this._filters.next({ ...this._filters.value, ...filters });
    }
}
