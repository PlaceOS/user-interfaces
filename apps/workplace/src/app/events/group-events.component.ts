import { Component } from '@angular/core';
import { GroupEventsStateService } from './group-events-state.service';
import { combineLatest } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { addDays, endOfDay, startOfDay } from 'date-fns';
import { AsyncHandler, SettingsService } from '@placeos/common';

@Component({
    selector: '[group-events]',
    template: `
        <topbar></topbar>
        <main class="flex flex-1 h-1/2 bg-base-200">
            <group-events-sidebar></group-events-sidebar>
            <div class="w-1/2 flex-1 h-full overflow-auto p-4">
                <group-events-filters-list></group-events-filters-list>
                <group-event-card
                    *ngIf="featured | async"
                    [event]="featured | async"
                    [featured]="true"
                    class="my-2 mx-auto w-[64rem] max-w-full"
                ></group-event-card>
                <ng-container
                    *ngIf="(event_list | async)?.length; else no_events"
                >
                    <div
                        class="flex flex-wrap mt-2 w-[64rem] max-w-full mx-auto"
                    >
                        <group-event-card
                            *ngFor="
                                let event of events_without_featured | async
                            "
                            [event]="event"
                            class="m-2"
                        ></group-event-card>
                    </div>
                </ng-container>
                <ng-template #no_events>
                    <div
                        class="flex flex-col items-center justify-center w-full h-full space-y-2"
                    >
                        <img src="assets/icons/no-results.svg" class="w-32" />
                        <div class="font-medium">
                            No events for the selected time period
                        </div>
                        <div class="opacity-30">
                            Expand you search or try again
                        </div>
                    </div>
                </ng-template>
            </div>
        </main>
        <footer-menu></footer-menu>
    `,
    styles: [
        `
            :host {
                position: absolute;
                top: 0;
                left: 0;
                right: 0;
                bottom: 0;
                overflow: hidden;
                display: flex;
                flex-direction: column;
            }

            main {
                min-height: 50%;
            }

            .top {
                background-color: #007ac8;
            }
        `,
    ],
})
export class GroupEventsComponent extends AsyncHandler {
    public readonly event_list = this._state.filtered_events;
    public readonly featured = this.event_list.pipe(
        map((_) => _.find((_: any) => _.extension_data?.featured || _.featured))
    );
    public readonly events_without_featured = combineLatest([
        this.event_list,
        this.featured,
    ]).pipe(
        map(([list, featured]) =>
            list.filter((_: any) => _.id !== featured?.id)
        )
    );

    constructor(private _state: GroupEventsStateService) {
        super();
    }
}
