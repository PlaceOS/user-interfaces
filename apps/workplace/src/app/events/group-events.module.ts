import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { RouterModule, Route } from '@angular/router';

import { FormFieldsModule } from '@placeos/form-fields';

import { GroupEventsComponent } from './group-events.component';
import { SharedComponentModule } from '../components/shared.module';
import { GroupEventsSidebarComponent } from './group-events-sidebar.component';
import { SharedBookingsModule } from '@placeos/bookings';
import { GroupEventsFiltersListComponent } from './group-events-filters-list.component';

const ROUTES: Route[] = [{ path: '', component: GroupEventsComponent }];

@NgModule({
    declarations: [
        GroupEventsComponent,
        GroupEventsSidebarComponent,
        GroupEventsFiltersListComponent,
    ],
    imports: [
        CommonModule,
        FormsModule,
        SharedComponentModule,
        SharedBookingsModule,
        FormFieldsModule,
        RouterModule.forChild(ROUTES),
    ],
})
export class GroupEventsModule {}
