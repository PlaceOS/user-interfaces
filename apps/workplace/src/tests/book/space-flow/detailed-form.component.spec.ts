import { ReactiveFormsModule } from '@angular/forms';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { CateringStateService } from '@placeos/catering';
import { CalendarEvent, generateEventForm } from '@placeos/events';
import {
    DateFieldComponent,
    DurationFieldComponent,
    TimeFieldComponent,
    UserListFieldComponent,
} from '@placeos/form-fields';
import { OrganisationService } from '@placeos/organisation';
import { DetailBookSpaceFormComponent } from 'apps/workplace/src/app/book/space-flow/detailed-form.component';
import { MockComponent } from 'ng-mocks';

describe('DetailBookSpaceFormComponent', () => {
    let spectator: Spectator<DetailBookSpaceFormComponent>;
    const createComponent = createComponentFactory({
        component: DetailBookSpaceFormComponent,
        declarations: [
            MockComponent(DateFieldComponent),
            MockComponent(TimeFieldComponent),
            MockComponent(DurationFieldComponent),
            MockComponent(UserListFieldComponent),
        ],
        providers: [
            { provide: OrganisationService, useValue: {} },
            { provide: CateringStateService, useValue: {} }
        ],
        imports: [ReactiveFormsModule, MatFormFieldModule, MatInputModule],
    });

    beforeEach(() => (spectator = createComponent()));

    it('should create component', () => {
        expect(spectator.component).toBeTruthy();
    });

    it('should match snapshot', () => {
        expect(spectator.element).toMatchSnapshot();
        spectator.setInput({ form: generateEventForm(new CalendarEvent()) });
        spectator.detectChanges();
        expect(spectator.element).toMatchSnapshot();
    });
});
