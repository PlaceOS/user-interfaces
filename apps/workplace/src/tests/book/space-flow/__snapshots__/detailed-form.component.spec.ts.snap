// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailBookSpaceFormComponent should match snapshot 1`] = `
<div
  id="root1"
>
  
</div>
`;

exports[`DetailBookSpaceFormComponent should match snapshot 2`] = `
<div
  id="root1"
>
  <form
    class="ng-untouched ng-pristine ng-invalid"
    novalidate=""
  >
    <section
      class="mb-4 border-b border-gray-300"
    >
      <div
        class="flex flex-col sm:flex-row space-x-0 sm:space-x-2 w-[640px] max-w-[calc(100%-2rem)] mx-auto"
      >
        <div
          class="flex flex-col flex-1 w-full sm:w-1/3"
        >
          <label>
            Date
          </label>
          <a-date-field
            class="ng-untouched ng-pristine ng-valid"
            formcontrolname="date"
          >
             Date and time must be in the future 
          </a-date-field>
        </div>
        <div
          class="flex flex-col flex-1 w-full sm:w-1/3"
        >
          <label>
            Start Time
          </label>
          <a-time-field
            class="ng-untouched ng-pristine ng-valid"
            formcontrolname="date"
          />
        </div>
        <div
          class="flex flex-col flex-1 w-full sm:w-1/3"
        >
          <label>
            End Time
          </label>
          <a-duration-field
            class="ng-untouched ng-pristine ng-valid"
            formcontrolname="duration"
          />
        </div>
      </div>
    </section>
    <section
      class="mb-4 border-b border-gray-300"
    >
      <div
        class="flex flex-col w-[640px] max-w-[calc(100%-2rem)] mx-auto"
      >
        <label>
          Attendees
        </label>
        <a-user-list-field
          class="ng-untouched ng-pristine ng-valid"
          formcontrolname="attendees"
        />
      </div>
    </section>
    <section
      class="mb-4 border-b border-gray-300"
    >
      <div
        class="w-[640px] max-w-[calc(100%-2rem)] mx-auto"
      >
        <div
          class="flex flex-col"
        >
          <label>
            Meeting Subject 
            <span>
              *
            </span>
          </label>
          <mat-form-field
            appearance="outline"
            class="mat-form-field mat-primary mat-form-field-type-mat-input mat-form-field-appearance-outline mat-form-field-can-float ng-untouched ng-pristine ng-invalid _mat-animation-noopable"
          >
            <div
              class="mat-form-field-wrapper"
            >
              <div
                class="mat-form-field-flex"
              >
                <div
                  class="mat-form-field-outline"
                >
                  <div
                    class="mat-form-field-outline-start"
                  />
                  <div
                    class="mat-form-field-outline-gap"
                  />
                  <div
                    class="mat-form-field-outline-end"
                  />
                </div>
                <div
                  class="mat-form-field-outline mat-form-field-outline-thick"
                >
                  <div
                    class="mat-form-field-outline-start"
                  />
                  <div
                    class="mat-form-field-outline-gap"
                  />
                  <div
                    class="mat-form-field-outline-end"
                  />
                </div>
                
                
                
                <div
                  class="mat-form-field-infix"
                >
                  <input
                    aria-invalid="false"
                    aria-required="false"
                    class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-invalid cdk-text-field-autofill-monitored"
                    data-placeholder="Meeting title"
                    formcontrolname="title"
                    matinput=""
                    placeholder="Meeting title"
                  />
                  <span
                    class="mat-form-field-label-wrapper"
                  >
                    
                  </span>
                </div>
                
              </div>
              
              <div
                class="mat-form-field-subscript-wrapper"
              >
                
                <div
                  class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
                  style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                >
                  
                  <div
                    class="mat-form-field-hint-spacer"
                  />
                </div>
                
              </div>
            </div>
          </mat-form-field>
        </div>
        <div
          class="flex flex-col resize-y"
        >
          <label>
            Notes
          </label>
          <mat-form-field
            appearance="outline"
            class="mat-form-field mat-primary mat-form-field-type-mat-input mat-form-field-appearance-outline mat-form-field-can-float ng-untouched ng-pristine ng-valid _mat-animation-noopable"
          >
            <div
              class="mat-form-field-wrapper"
            >
              <div
                class="mat-form-field-flex"
              >
                <div
                  class="mat-form-field-outline"
                >
                  <div
                    class="mat-form-field-outline-start"
                  />
                  <div
                    class="mat-form-field-outline-gap"
                  />
                  <div
                    class="mat-form-field-outline-end"
                  />
                </div>
                <div
                  class="mat-form-field-outline mat-form-field-outline-thick"
                >
                  <div
                    class="mat-form-field-outline-start"
                  />
                  <div
                    class="mat-form-field-outline-gap"
                  />
                  <div
                    class="mat-form-field-outline-end"
                  />
                </div>
                
                
                
                <div
                  class="mat-form-field-infix"
                >
                  <textarea
                    aria-invalid="false"
                    aria-required="false"
                    class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-valid cdk-text-field-autofill-monitored"
                    data-placeholder="Add meeting notes here..."
                    formcontrolname="body"
                    matinput=""
                    placeholder="Add meeting notes here..."
                  />
                  <span
                    class="mat-form-field-label-wrapper"
                  >
                    
                  </span>
                </div>
                
              </div>
              
              <div
                class="mat-form-field-subscript-wrapper"
              >
                
                <div
                  class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
                  style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                >
                  
                  <div
                    class="mat-form-field-hint-spacer"
                  />
                </div>
                
              </div>
            </div>
          </mat-form-field>
        </div>
        
      </div>
    </section>
  </form>
  
</div>
`;
